include/vcpkg/registries.h

```diff
@@ -178,31 +177,8 @@ namespace vcpkg
         Filesystem,
     };
 
-    struct VersionDbEntryDeserializer final : Json::IDeserializer<VersionDbEntry>
-    {
-        static constexpr StringLiteral GIT_TREE = "git-tree";
-        static constexpr StringLiteral PATH = "path";
-
-        LocalizedString type_name() const override;
-        View<StringView> valid_fields() const override;
-        Optional<VersionDbEntry> visit_object(Json::Reader& r, const Json::Object& obj) const override;
-        VersionDbEntryDeserializer(VersionDbType type, const Path& root) : type(type), registry_root(root) { }
-
-    private:
-        VersionDbType type;
-        Path registry_root;
-    };
-
-    struct VersionDbEntryArrayDeserializer final : Json::IDeserializer<std::vector<VersionDbEntry>>
-    {
-        virtual LocalizedString type_name() const override;
-        virtual Optional<std::vector<VersionDbEntry>> visit_array(Json::Reader& r,
-                                                                  const Json::Array& arr) const override;
-        VersionDbEntryArrayDeserializer(VersionDbType type, const Path& root) : underlying{type, root} { }
-
-    private:
-        VersionDbEntryDeserializer underlying;
-    };
+    std::unique_ptr<Json::IDeserializer<std::vector<VersionDbEntry>>> make_version_db_deserializer(VersionDbType type,
```

Comment:
I'm not sure I agree with this one. It looks like the intent was to avoid naming JSON things in this interface but I'm not sure it actually does that because `IDeserializer` is named here. Notably, being in a `unique_ptr` means that `IDeserializer` will need to be complete for all callers (since `unique_ptr::~unique_ptr` does `T->~T()` which requires that T be complete).

If the problem is that this drags unrelated json stuff into unrelated TUs I would rather see a specific header for that rather than ptring it for that reason.

(If there is something other than trying to avoid dragging in the json stuff I missed then this might be fine. I don't object to unique_ptring on principle but I prefer solutions that avoid potential incomplete type diagnostic message hell)

Reply:
This function is only exposed for testing purposes -- 99% of the TUs including this header do not call it. Therefore, those TUs should only be aware that it is a valid type.

The reason it's a `unique_ptr` is to further separate the implementation from the interface -- the testing code should not need even the name of the deserializer implementation type. I'd strongly prefer this to return a reference to an `::instance`, but these deserializers haven't been made stateless :(.